satpy/readers/yaml_reader.py

```diff
@@ -516,59 +512,69 @@ def create_filehandlers(self, filenames, fh_kwargs=None):
                     self.file_handlers.get(filetype, []) + filehandlers,
                     key=lambda fhd: (fhd.start_time, fhd.filename))
 
-        # update existing dataset IDs with information from the file handler
-        self.update_ds_ids_from_file_handlers()
-
         # load any additional dataset IDs determined dynamically from the file
-        self.add_ds_ids_from_files()
+        # and update any missing metadata that only the file knows
+        self.update_ds_ids_from_file_handlers()
         return created_fhs
 
-    def update_ds_ids_from_file_handlers(self):
-        """Update DatasetIDs with information from loaded files.
+    def _file_handlers_available_datasets(self):
+        """Generate a series of available dataset information.
+
+        This is done by chaining file handler's
+        :meth:`satpy.readers.file_handlers.BaseFileHandler.available_datasets`
+        together. See that method's documentation for more information.
 
-        This is useful, for example, if dataset resolution may change
-        depending on what files were loaded.
+        Returns:
+            Generator of (bool, dict) where the boolean tells whether the
+            current dataset is available from any of the file handlers. The
+            boolean can also be None in the case where no loaded file handler
+            is configured to load the dataset. The
+            dictionary is the metadata provided either by the YAML
+            configuration files or by the file handler itself if it is a new
+            dataset. The file handler may have also supplemented or modified
+            the information.
 
         """
-        for file_handlers in self.file_handlers.values():
-            fh = file_handlers[0]
-            # update resolution in the dataset IDs for this files resolution
-            res = getattr(fh, 'resolution', None)
-            if res is None:
-                continue
+        # flatten all file handlers in to one list
+        flat_fhs = (fh for fhs in self.file_handlers.values() for fh in fhs)
```

Comment:
- If there are two filehandlers with different resolutions, only the first one gets to update the dataset info, because the second one yields early, right?
- Is there a way to avoid checking all file handlers if we already know that one is sufficient (e.g. for SEVIRI HRIT there are 10 segments per channel)

Reply:
For your first point and your questions about missing continues, no. There should not be continues so that the current file handler can continue on and add a new resolution if it needs to. That's why it needs to check to make sure the current ds_info's resolution is different than the resolution it wants to add. At least I think this is where you were going.

For your second point, this is a good point. I originally only had the first file handler searched but @mraspaud pointed out that that wouldn't work for VIIRS SDR. I think any checks in the SEVIRI HRIT file handlers (if custom) need to be careful on how they do things.

Reply:
For SEVIRI HRIT, the file name tells us already the available dataset, so the check should be really quick. Also, each file contains only one dataset, so we can return as soon as we find something

Reply:
1. Ok, I see. A new dataset with a different resolution is yielded then. Thanks for the clarification.
2. Ok, good!